### PR TITLE
feat(connector): [Payeezy] Currency Unit Conversion

### LINKED ISSUE #2241: [FEATURE]: [Payeezy] Currency Unit Conversion
### :memo: Feature Description

- Each currency can be described in terms of base or minor units. 
- For instance, when dealing with USD currency, the Cent is considered as the minor unit, while the Dollar is considered the base unit. 
- In Hyperswitch, the amount value is expected to be always provided in minor units. 
- For example: In USD, If the amount is 1 in base unit (Dollar), then it will be equivalent to 100 in minor units (Cent)
- Some of the connectors integrated require the amount to be converted to a desirable unit before being passed to them. 
- We have functions `to_currency_base_unit` and `to_currency_lower_unit` in place to convert the minor unit amount to its decimal equivalent and vice versa. These conversions are handled explicitly based on the connector.

### :hammer: Possible Implementation

- ConnectorCommon trait have been implemented for the connector. 
- This trait contains `get_currency_unit` method. This method needs to be implemented. 
- It will define what type of conversion needs to be done during `connector_router_data` creation. Concurrently handle the creation of `connector_router_data` to be passed for the connector request body creation.
- You can check this PR for further reference https://github.com/juspay/hyperswitch/pull/2196

:bookmark: Note: All the changes needed should be contained within `hyperswitch/crates/router/src/connector/`

### :package: Have you spent some time checking if this feature request has been raised before?

- [X] I checked and didn't find a similar issue

### :package: Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### PR DESCRIPTION
**Pull request by adarsh-jha-dev**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

Modified two files in `hyperswitch/crates/router/src/connector/`

- Addressing Issue #2241 
- `payeezy.rs`
    - Implement `get_currency_unit` function
    - Modify `ConnectorIntegration` implementations for `Payeezy`
- `payeezy/transformers.rs`
    - Implement `PayeezyRouterData<T>` structure and functionality
    - Modify implementation for `PayeezyPaymentsRequest`

## Motivation and Context

Fixes #2241 

## How did you test it?

We need to create a payment using payeezy and test the currency unit from connector response and connector dashboard.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by adarsh-jha-dev** [comment]
I request any of the maintainer to please have a look at the PR and let me know if the changes are relevant.

**Comment by adarsh-jha-dev** [comment]
> Hey @adarsh-jha-dev , Pls address the CI fails Run `cargo +nightly` fmt for formatting.

Hey @swangi-kumari , I am doing these changes on github codespaces due to some unavoidable reasons and it is becoming problematic for me to run the command for formatting. Is there any way you can run that command for me (it's a request) and if not then please give me a little bit of time to do the same.

**Comment by adarsh-jha-dev** [comment]
> Hey @adarsh-jha-dev , Pls address the CI fails Run `cargo +nightly` fmt for formatting.

Done

**Comment by adarsh-jha-dev** [comment]
@SamraatBansal , I request you to please have a look at it and let me know if the changes are relevant?

**Comment by swangi-kumari** [comment]
Hey @adarsh-jha-dev ,
Please address the failing CI checks
Run `cargo clippy` and addressed lints thrown by it.

**Comment by adarsh-jha-dev** [comment]
> Hey @adarsh-jha-dev , Please address the failing CI checks Run `cargo clippy` and addressed lints thrown by it.

Done, could you please review this .

**Comment by swangi-kumari** [comment]
Hey @adarsh-jha-dev ,
Please address the failing CI checks
Run `cargo clippy` and addressed lints thrown by it.

**Comment by swangi-kumari** [comment]
Hey @adarsh-jha-dev ,
Your PR still have some compilation errors. 
Please address the failing CI checks
Run cargo clippy and addressed lints thrown by it.

**Comment by adarsh-jha-dev** [comment]
> Hey @adarsh-jha-dev , Your PR still have some compilation errors. Please address the failing CI checks Run cargo clippy and addressed lints thrown by it.

Yeah yeah , on it, would resolve it within an hour

**Comment by adarsh-jha-dev** [comment]
Hey @swangi-kumari, sorry for not submitting this yesterday, trying my best to submit by tonight. I request you to please don't unassign me

**Comment by swangi-kumari** [comment]
Sure @adarsh-jha-dev , Take your time and address the comments given to you. All the Best:)

**Comment by adarsh-jha-dev** [comment]
Thanks @swangi-kumari for co-operating and giving me time, I have tried hard and resolved those CI fails. I request you to please have a look at them whenever you want.

File: crates/router/src/connector/payeezy/transformers.rs
Comment (swangi-kumari) [inline crates/router/src/connector/payeezy/transformers.rs:35]:
```suggestion
    ) -> Result<Self, Self::Error> {
            let amount = utils::get_amount_as_string(currency_unit, amount, currency)?;
```

**Review by swangi-kumari — CHANGES_REQUESTED** [review]

**Comment by swangi-kumari** [comment]
Hey @adarsh-jha-dev ,
Please make this changes too.
line (95-104)
```
#[derive(Serialize, Debug)]
pub struct PayeezyPaymentsRequest {
    pub merchant_ref: String,
    pub transaction_type: PayeezyTransactionType,
    pub method: PayeezyPaymentMethodType,
    pub amount: String,
    pub currency_code: String,
    pub credit_card: PayeezyPaymentMethod,
    pub stored_credentials: Option<StoredCredentials>,
    pub reference: String,
}
```
Here in this struct convert amount to String 
 `pub amount: String,`

File: crates/router/src/connector/payeezy/transformers.rs
Comment (adarsh-jha-dev) [reply crates/router/src/connector/payeezy/transformers.rs:35]: Done , thanks for your assistance

File: crates/router/src/connector/payeezy.rs
Comment (AkshayaFoiger) [inline crates/router/src/connector/payeezy.rs:1]:
```
impl ConnectorIntegration<api::Capture, types::PaymentsCaptureData, types::PaymentsResponseData>
    for Payeezy
{
....
fn get_request_body(
        &self,
        req: &types::PaymentsCaptureRouterData,
    ) -> CustomResult<Option<types::RequestBody>, errors::ConnectorError> {
        let connector_req = payeezy::PayeezyCaptureOrVoidRequest::try_from(req)?;
        let payeezy_req = types::RequestBody::log_and_get_request_body(
            &connector_req,
            utils::Encode::<payeezy::PayeezyCaptureOrVoidRequest>::encode_to_string_of_json,
        )
        .change_context(errors::ConnectorError::RequestEncodingFailed)?;
        Ok(Some(payeezy_req))
}
}
```

In the above function for `PaymentsCaptureData`, do simlilar changes  as you did for `types::PaymentsAuthorizeRouterData`  and `types::RefundsRouterData<api::Execute>`  and address the corresponding changes required in `payeezy/transformers.rs` file

File: crates/router/src/connector/payeezy/transformers.rs
Comment (AkshayaFoiger) [inline crates/router/src/connector/payeezy/transformers.rs:490]:
```suggestion
            amount: item.amount,
```

**Review by AkshayaFoiger — CHANGES_REQUESTED** [review]

**Comment by swangi-kumari** [comment]
Hey @adarsh-jha-dev ,
Thanks for your interest in contributing to hyperswitch.
Let us know if you need any assistance from our end.
Also, even if hacktoberfest is over, we should celebrate open source everyday and we are open for more contributions from you.
We would still be rewarding folks with goodies even if the PR gets merged post hacktoberfest.
May the Source be with you!

**Comment by adarsh-jha-dev** [comment]
> Hey @adarsh-jha-dev , Thanks for your interest in contributing to hyperswitch. Let us know if you need any assistance from our end. Also, even if hacktoberfest is over, we should celebrate open source everyday and we are open for more contributions from you. We would still be rewarding folks with goodies even if the PR gets merged post hacktoberfest. May the Source be with you!

Yeah, I am almost done with the changes, I am happy to contribute to Hyperswitch even after the fest, it helped me learn a lot of things

File: crates/router/src/connector/payeezy.rs
Comment (adarsh-jha-dev) [reply crates/router/src/connector/payeezy.rs:1]: Done with the changes, could you please let me know if there are further changes required.

**Comment by adarsh-jha-dev** [comment]
Hey @swangi-kumari , I have done the changes requested by you and other maintainers, I request you to please proceed further.

File: crates/router/src/connector/payeezy/transformers.rs
Comment (AkshayaFoiger) [inline crates/router/src/connector/payeezy/transformers.rs:354]:
```suggestion
            amount: item.amount.clone(),
```

**Review by AkshayaFoiger — CHANGES_REQUESTED** [review]

**Comment by AkshayaFoiger** [comment]
Thanks for your quick response @adarsh-jha-dev,
Looks good to me!

**Review by AkshayaFoiger — APPROVED** [review]

**Comment by adarsh-jha-dev** [comment]
Hey @swangi-kumari , I request you to please have a look at it, whenever you're free

**Review by swangi-kumari — APPROVED** [review]

**Comment by adarsh-jha-dev** [comment]
@SamraatBansal  and @Sangamesh26 , could you also please approve this or let me know if there's any change needed?

**Comment by adarsh-jha-dev** [comment]
Hey @swangi-kumari , could you please merge this if you're satisfied with the changes?

**Review by SanchithHegde — APPROVED** [review]
Looks good to me!

Thanks for the PR, @adarsh-jha-dev!

**Pull request merged**
